### PR TITLE
schema: add JSON schema for mcollective, migrator and mounts modules            

### PR DESCRIPTION
**Pull request by blackboxsw**
## Proposed Commit Message
```
schema: add JSON schema for mcollective, migrator and mounts modules

Define static JSON schema for cloud-config modules in
cloudinit/config/cloud-init-schema.json. Migrate existing doc/examples
into inline meta schema examples where applicable.

Includes:
 - cc_mcollective
 - cc_migrator
 - cc_mounts

Doc drops:
 - doc/examples/cloud-config-mcollective.txt because example is included in
   inline meta schema docs
```

## Test Steps
```
tox -re doc; xdg-open doc/rtd_html/topics/modules.html 
# validate visual changes vs https://cloudinit.readthedocs.io/en/latest/topics/modules.html#mcollective

# validate specific module rendered docs on cmdline
PYTHONPATH=. python3 -m cloudinit.cmd.main devel schema --docs cc_mcollective
```
Unit tests will run through any included cc_<module>.meta['examples'] to confirm that they are correct schema. 

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

### CODE REVIEW

File: cloudinit/config/cloud-init-schema.json [hunk 1]
```diff
@@ -783,6 +783,97 @@
...
+                "public-cert": {
+                  "type": "string",
+                  "description": "Optional value of server public certificate which will be written to ``/etc/mcollective/ssl/server-public.pem``"
+                },
+                "private-cert": {
+                  "type": "string",
+                  "description": "Optional value of server private certificate which will be written to ``/etc/mcollective/ssl/server-private.pem``"
+                }
+              },
+              "patternProperties": {
```
Comment (TheRealFalcon) [inline cloudinit/config/cloud-init-schema.json:804]: Do `properties` and `patternProperties` co-exist at the same level peacefully? Can we add a test where we set public-cert or private-cert to a boolean or integer and verify it fails to show that the `patternProperties` definition isn't replacing the `properties` definition?

File: cloudinit/config/cloud-init-schema.json [hunk 2]
```diff
@@ -783,6 +783,97 @@
...
+      "properties": {
+        "mounts": {
+          "type": "array",
+          "items": {
+              "type": "array",
+              "items": {"type": "string"},
+              "minItems": 1,
+              "maxItems": 6
+          },
+          "description": "List of lists. Each inner list entry is a list of ``/etc/fstab`` entries of the format: [ fs_spec, fs_file, fs_vfstype, fs_mntops, fs-freq, fs_passno ]",
```
Comment (TheRealFalcon) [inline cloudinit/config/cloud-init-schema.json:841]: We should have some mention that all 6 entries are not required.

File: cloudinit/config/cc_mounts.py [hunk 3]
```diff
@@ -42,32 +59,35 @@
...
-        maxsize: <size in bytes>
-"""
-
-import logging
-import os
-import re
-from string import whitespace
-
-from cloudinit import subp, type_utils, util
+        filename: /my/swapfile
```
Comment (TheRealFalcon) [inline cloudinit/config/cc_mounts.py:72]: In lines 66-72, is it worth removing quotes around options that don't need it (or possibly putting quotes around everything?). I'm just noticing the inconsistency between `auto` and `"auto"`

File: cloudinit/config/cc_mounts.py [hunk 4]
```diff
@@ -42,32 +59,35 @@
 with ``filename``, the size of the swap file with ``size`` maximum size of
 the swap file if using an ``size: auto`` with ``maxsize``. By default no
 swap file is created.
+"""
```
Comment (TheRealFalcon) [inline cloudinit/config/cc_mounts.py:62]: We should update the blurb mentioning the default:
```
It defaults to::
    mount_default_fields: [none, none, "auto", "defaults,nobootwait", "0", "2"]
On a systemd booted system that default is the mostly equivalent::
    mount_default_fields: [none, none, "auto",
       "defaults,nofail,x-systemd.requires=cloud-init.service", "0", "2"]
Note that `nobootwait` is an upstart specific boot option that somewhat
equates to the more standard `nofail`.
```
Is upstart default anywhere anymore? I think we should just specify the systemd default as the only default and then mention that other init systems may vary in their defaults.

File: cloudinit/config/cloud-init-schema.json [hunk 5]
```diff
@@ -783,6 +783,97 @@
...
+          "minItems": 1
+        },
+        "mount_default_fields": {
+          "type": "array",
+          "description": "Default mount configuration for any mount entry with less than 6 options provided. When specified, 6 items are required and represent ``/etc/fstab`` entries. Systemd-based mount options will be ``defaults,nofail,x-systemd.requires=cloud-init.service,_netdev``",
+          "default": [null, null, "auto", "defaults,nobootwait", "0", "2"],
```
Comment (TheRealFalcon) [inline cloudinit/config/cloud-init-schema.json:847]: See my other comment about defaults. I think we should use the systemd default here and not specify other defaults.

File: cloudinit/config/cloud-init-schema.json [hunk 6]
```diff
@@ -783,6 +783,97 @@
...
+        "swap": {
+          "type": "object",
+          "properties": {
+            "filename": {
+              "type": "string",
+              "description": "Path to the swap file to create"
+            },
+            "size": {
+              "description": "The size in bytes of the swap file, or 'auto'",
+              "oneOf": [{"enum": ["auto"]}, {"type": "integer"}]
```
Comment (TheRealFalcon) [inline cloudinit/config/cloud-init-schema.json:866]: String is valid here: https://github.com/canonical/cloud-init/blob/main/cloudinit/config/cc_mounts.py#L378-L381

File: cloudinit/config/cloud-init-schema.json [hunk 7]
```diff
@@ -783,6 +783,97 @@
...
+            },
+            "maxsize": {
+              "type": "integer",
```
Comment (TheRealFalcon) [inline cloudinit/config/cloud-init-schema.json:869]: String is valid here: https://github.com/canonical/cloud-init/blob/main/cloudinit/config/cc_mounts.py#L378-L381

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
Some comments inline, but overall looking good.

File: cloudinit/config/cloud-init-schema.json (hunk 1 above)
Comment (blackboxsw) [reply cloudinit/config/cloud-init-schema.json:804]: Yes they do. I had wondered the same things, which is why we [already have a test for explicit property types like public-cert in the unittests of this PR](https://github.com/canonical/cloud-init/pull/1358/files#diff-226b092ec1d5779ac38368bed2be8bb6a67701d48155e01604039169005906c2R176-R178)

File: cloudinit/config/cloud-init-schema.json (hunk 6 above)
Comment (blackboxsw) [reply cloudinit/config/cloud-init-schema.json]: added `"pattern": "^([0-9]+)?\\.?[0-9]+[BKMGT]$"}` to schema for swap.size swap.maxsize and unittests

File: cloudinit/config/cloud-init-schema.json [hunk 8]
```diff
@@ -838,13 +838,13 @@
               "minItems": 1,
               "maxItems": 6
           },
-          "description": "List of lists. Each inner list entry is a list of ``/etc/fstab`` entries of the format: [ fs_spec, fs_file, fs_vfstype, fs_mntops, fs-freq, fs_passno ]",
+          "description": "List of lists. Each inner list entry is a list of ``/etc/fstab`` mount declaratiosn of the format: [ fs_spec, fs_file, fs_vfstype, fs_mntops, fs-freq, fs_passno ]. A mount declaration with less than 6 items, will get remaining values from ``mount_default_fields``. A mount declaration with only `fs_spec` and no `fs_file` mountpoint will be skipped.",
```
Comment (TheRealFalcon) [inline cloudinit/config/cloud-init-schema.json:841]: `declaratiosn`->`declarations`. Also the comma in the second sentence should be removed.

File: cloudinit/config/cc_mounts.py [hunk 9]
```diff
@@ -45,15 +46,10 @@
...
-
-On a systemd booted system that default is the mostly equivalent::
-
     mount_default_fields: [none, none, "auto",
-       "defaults,nofail,x-systemd.requires=cloud-init.service", "0", "2"]
+        "defaults,nofail,x-systemd.requires=cloud-init.service", "0", "2"]
 
-Note that `nobootwait` is an upstart specific boot option that somewhat
-equates to the more standard `nofail`.
+Other init systems will vary in ``mount_default_fields``.
```
Comment (TheRealFalcon) [inline cloudinit/config/cc_mounts.py:52]: Nit: might be clearer to change `Other` to `Non-systemd`

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
One typo we should fix, and everything else optional. For some reason Github wouldn't let me do suggestions on this review.

**Pull request merged**